### ISSUE TITLE
Bug: support `as const` on `RuleLevelAndOptions`

### ISSUE THREAD

**Issue by abrahamguo** [typescript-eslint/typescript-eslint#10188]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

```TypeScript
import tseslint from 'typescript-eslint';

const curly = ['error'] as const;
tseslint.config({ rules: { curly } });
```

### tsconfig

```jsonc
{ "compilerOptions": { "module": "nodenext" } }
```

### Expected Result

The types should allow using `as const` without any type errors.

### Actual Result

TS error:
```
Type 'readonly ["error"]' is not assignable to type 'RuleEntry | undefined'.
  The type 'readonly ["error"]' is 'readonly' and cannot be assigned to the mutable type 'RuleLevelAndOptions'.
```

### Additional Info

Of course, removing `as const` does not work because of the tuple type, but this is expected —
```
Type 'string[]' is not assignable to type 'RuleEntry | undefined'.
  Type 'string[]' is not assignable to type 'RuleLevelAndOptions'.
    Source provides no match for required element at position 0 in target.
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/utils`             | `8.10.0` |
| `TypeScript`                           | `5.6.3` |

**Comment by bradzacher** [comment]
Why are you configuring core rules with separate variables...? This style will ONLY work with the core rules - all plugins have slash names so you can't use the shorthand syntax.

**Comment by abrahamguo** [comment]
Sure thing — my example above was a minimal reproducible example to demonstrate the issue.

Here's an example more like what I'm doing:
```
import tseslint from 'typescript-eslint';

const rules = Object.fromEntries(
	Object.entries({
		'consistent-type-imports': { prefer: 'no-type-imports' },
		'consistent-type-definitions': 'type',
		'no-unused-expressions': { enforceForJSX: true }
		// ... many other rules and other plugins ...
	}).map(([ruleName, options]) => [
		`@typescript-eslint/${ruleName}`,
		['error', options]
	])
);

export default tseslint.config({
	plugins: { '@typescript-eslint': tseslint.plugin },
	rules
});
```

Basically, I have a large object with custom options, that I'm wanting to dynamically add the `'error'` level to.

**Comment by JoshuaKGoldberg** [comment]
> since the convention for tuple types throughout the rest of the TS ecosystem is generally that you can either use `as const` or a type annotation

I don't think that's always true. Readonly types (`as const`) are pretty restrictive; most projects aren't great about consistently allowing readonly parameters in APIs. We're not very different from the norm.

Also, a `satisfies` is common too. I think in your case the current canonically reasonable thing would be `satisfies tseslint.RuleLevelAndOptions` or some such.

**Comment by bradzacher** [comment]
Yeah I would love to do this but adding a readonly to the types can be quite terrible for end users. If you union it in then no one can read the values. If you add it in wholesale then the same thing occurs.

TS just sucks here with its handling of tuples.

**Comment by bradzacher** [comment]
With 2 -1's I'm going to close this.
